[macro] fix redefinition of modules with defineType/defineModule

kLabz opened the pull request:
Fixes issues like `HxbData.HxbFailure("Could not read static field test on Foo while hxbing Main")` on current nightlies.

* [x] Implement `haxe.macro.CompilationServer.invalidateModule()` (I kept that, as it can be useful in some cases)
* [x] Rework the implementation: error when redefining a loaded module, invalidate previous module when redefining a module that is not loaded yet
* [x] Clean up tests; some don't make much sense now / are redundant

Follow-up:
* [ ] 4.3_bugfix PR with a fix to the related issue of defineType/defineModule modules having an empty file and being immediately invalidated

kLabz committed:
[macro] Disallow redefining module through Context's defineType/defineModule

kLabz committed:
[macro] defineType / defineModule: allow redefining MSBad modules

Note: this hook should be moved somewhere else..

kLabz committed:
[tests] small cleanup

kLabz committed:
[tests] add test

kLabz commented on src/typing/typeloadCacheHook.ml:
```diff
@@ -10,6 +10,7 @@ type find_module_result =
 	| BinaryModule of HxbData.module_cache
 	| NoModule
 
+let find_module_extra_hook : (Common.context -> path -> module_def_extra) ref = ref (fun _ _ -> raise Not_found)
```
Note: this is only here temporarily, it should be moved to a better suited place

kLabz committed:
[tests] add more tests

kLabz committed:
[std] add haxe.macro.CompilationServer.invalidateModule()

kLabz committed:
Fix defineModule into curmod

kLabz committed:
[tests] add test using CompilationServer.invalidateModule

kLabz committed:
Default impl without server

kLabz committed:
[macro] disallow invalidating loaded module

kLabz committed:
CompilationServer.invalidate: raise error that macros can catch; add test

Simn commented:
This whole hook approach doesn't make sense to me. There should be exactly two cases when dealing with modules via `define_module` and `define_type`:

1. The module already exists in the current compilation context (`com.module_lut`). In this case we give a redefinition error.
2. The module doesn't exist in the current compilation context. In this case we add it and dirty any cached modules matching it. This will cause any of its dependents to pick up the new module.

What exactly isn't working with this general approach?

kLabz committed:
Rework implementation

Allow redefining when not loaded, but invalidate previous cached module (if any)

kLabz committed:
[tests] update tests

kLabz committed:
Use more specific invalidation reasons

kLabz committed:
[tests] update tests

kLabz commented:
> This whole hook approach doesn't make sense to me. There should be exactly two cases when dealing with modules via `define_module` and `define_type`:
> 
> 1. The module already exists in the current compilation context (`com.module_lut`). In this case we give a redefinition error.
> 2. The module doesn't exist in the current compilation context. In this case we add it and dirty any cached modules matching it. This will cause any of its dependents to pick up the new module.
> 
> What exactly isn't working with this general approach?

This seems fine indeed. I updated the PR and the tests (which fail with HxbFailure on dev).

kLabz committed:
Let macros catch redefinition errors

kLabz committed:
[tests] test catching errors

Simn commented on std/haxe/macro/CompilationServer.hx:
```diff
@@ -69,6 +69,16 @@ class CompilationServer {
+		Invalidates a module, removing it from the cache.
+
+		If the module has already been loaded in current context, a compiler
+		error will be raised which can be caught using `try ... catch`.
```
This should probably mention what to actually catch.

kLabz replied:
I'm not against it, but we also don't mention it on other cases (`Context` API for example).

Simn commented on std/haxe/macro/CompilationServer.hx:
```diff
@@ -69,6 +69,16 @@ class CompilationServer {
+		If the module has already been loaded in current context, a compiler
+		error will be raised which can be caught using `try ... catch`.
+	**/
+	static public function invalidateModule(path:String) {
```
In general I'm somewhat worried about exposing this to users because it requires some knowledge of how our compilation cache works. Could you tell me in which cases this function is supposed to be used?

kLabz replied:
Most recent one I've seen is a dependency on an external file but only for specific changes (for example a field in a json file, with the file itself changing all the time)

Simn replied:
That sounds pretty wild... but well, I guess I'll not question it.

kLabz committed:
[std] more information about catching errors
